cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use anyhow::{anyhow, Result};
+use crc_any::CRCu16;
+use num_derive::FromPrimitive;
+//use packed_struct::prelude::*;
+use std::cmp::min;
+use std::convert::TryInto;
+use strum_macros::EnumString;
+use zerocopy::{AsBytes, FromBytes};
+
+#[repr(u8)]
+#[derive(Debug)]
+enum PacketType {
+    Ack = 0xA1,
+    //Nak = 0xA2,
+    AckAbort = 0xA3,
+    Command = 0xA4,
+    Data = 0xA5,
+    Ping = 0xA6,
+    PingResponse = 0xA7,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug)]
+pub enum ResponseCode {
+    Generic = 0xA0,
+    ReadMemory = 0xA3,
+    GetProperty = 0xA7,
+    //FlashReadOnce = 0xAF,
+    KeyProvision = 0xB5,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum KeyType {
+    // Secureboot Key Encryption Key
+    // SBKEK = 0x3,
+    // Prince = 7 - 9
+    // USER is available to use for whatever
+    // Wish there were more than one user :(
+    // USERKEK = 0xB,
+    // UDS used in DICE
+    UDS = 0xC,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum KeyProvisionCmds {
+    Enroll = 0x0,
+    //SetUserKey = 0x1,
+    SetIntrinsicKey = 0x2,
+    WriteNonVolatile = 0x3,
+    //ReadNonVolatile = 0x4,
+    WriteKeyStore = 0x5,
+    //ReadKeyStore = 0x6,
+}
+
+// Commands are abbridged right now for what we care about
+#[repr(u8)]
+#[derive(Debug)]
+pub enum CommandTag {
+    FlashEraseAll = 0x1,
+    ReadMemory = 0x3,
+    WriteMemory = 0x4,
+    GetProperty = 0x7,
+    KeyProvision = 0x15,
+}
+
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes)]
+pub struct PacketHeader {
+    start_byte: u8,
+    packet_type: u8,
+}
+
+impl PacketHeader {
+    fn new(ptype: PacketType) -> PacketHeader {
+        PacketHeader { start_byte: 0x5A_u8, packet_type: ptype as u8 }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, EnumString, FromPrimitive, Clone, Copy)]
+pub enum BootloaderProperty {
+    BootloaderVersion = 1,
+    AvailablePeripherals = 2,
+    FlashStart = 3,
+    FlashSize = 4,
+    FlashSectorSize = 5,
+    AvailableCommands = 7,
+    CRCStatus = 8,
+    VerifyWrites = 10,
+    MaxPacketSize = 11,
+    ReservedRegions = 12,
+    RAMStart = 14,
+    RAMSize = 15,
+    SystemDeviceID = 16,
+    SecurityState = 17,
+    UniqueID = 18,
+    TargetVersion = 24,
+    FlashPageSize = 27,
+    IRQPinStatus = 28,
+    FFRKeyStoreStatus = 29,
+}
+
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes)]
+pub struct PingResponse {
+    header: PacketHeader,
+    protocol_bugfix: u8,
+    protocol_minor: u8,
+    protocol_major: u8,
+    protocol_name: u8,
+    options_low: u8,
```

Comment:
You could use `zerocopy::byteorder::U16<BigEndian>` (or is it `LittleEndian`?) for all of the two-byte values in the various `struct`s, which would save manually building them in constructors!